### DISCUSSION TITLE
Local Coop Showcase

### DISCUSSION THREAD

**Discussion by RisingLeaf** [endless-sky/endless-sky#9497] [Show and tell]
## Intro
Endless Sky is fun, a lot of fun. So naturally we strive to make it more fun and what is more fun than shooting pirate ships and flying through unknown territory? Doing it with a friend of course, shared fun is double fun!

Presenting a first proto type of a local coop with split-screen:
https://github.com/RisingLeaf/endless-sky/tree/local-coop

## Features
Not a lot, this is mostly a prove of concept, so I ran with the minimum viable product(still 1000 lines changed though :sweat_smile:).
But here is a quick overview:
- Splitscreen
- Control over the ship that is last in the list of player ships(limited, some controls like afterburners are still missing)
- The escort will automatically jump when the player does/it will land on wormholes
- pages for the controls as the extra controls were not fitting on the one page that was there.

The split screen is always on, I was too lazy to make it possible to toggle yet.

## Technical details
The split screen is created with framebuffers, creating a duplicate of each draw list. This is not doubling gpu load though, it is increasing it, but because the decreased draw size it is also nearly halving the load per screen, so I experienced an increase of about 1.5x.
 I am still experience some problems with resizing the window but that should be fixable.

## Additional Notes
I could imagine this as a feature for main ES, but the code is really ugly currently and it is far from complete so I wont PR it anytime soon if ever, I am opening this discussion to show it is possible and to just open a space for suggestions and if people want to use it they can clone the branch and play with it.
It is also for the devs to decide if this is even something that is wanted for the game.

## Related stuff
There is actually a bunch of related issues:
#8569
#327
#266
#149
#1705
Not all of them are talking about a local coop, but I thought linking them here wont hurt.

## Screenrecording

https://github.com/endless-sky/endless-sky/assets/85687254/f0351825-7dd0-4bf4-8d9d-cb7625d5640f

(I know the quality is bad)

**Comment by RisingLeaf** [comment]
UPDATE:
-----------------------
Split screen is now toggle-able with a user defined key.

Still noticing some issues, but it is getting better (we are now at  1,132 additions and 385 deletions).

I might want to use the now existing framebuffer solution but hey can not be rendered with the sprite shader, so I think I am just not going to do it :stuck_out_tongue:

**Comment by Bubadark** [comment]
as a almost entirely non Visual Studio user what would it take to have someone do a build for this that i can play around with? Love what i see Rising Leaf! maybe if i am to get a bit more involved i should learn, craped out trying to build it my first few try's though.

**Reply by RisingLeaf** [reply]
Refer to this document on how to build code: https://github.com/endless-sky/endless-sky/blob/master/docs/readme-developer.md

To get the code for this specific branch go here: https://github.com/RisingLeaf/endless-sky/tree/local-coop

and click on the code button to download the files as seen in the picture

<img width="930" alt="Screenshot 2024-03-04 at 7 43 55 PM" src="https://github.com/endless-sky/endless-sky/assets/85687254/f1f332d0-7d47-4cf0-825c-d17fb54beda7">
